### DISCUSSION TITLE
php Imagick exportImagePixels use a lot of memory

### DISCUSSION THREAD

**Discussion by retsetB** [ImageMagick/ImageMagick#5815] [Help]
Hi Everyone
I have a big image 3000x2000 this image use all 4 channels so this is RGBA and I need to process each pixels in some way.

I try to use function exportImagePixels and saw some crash.

in php I use following code:
`$lines = $this->image->exportImagePixels(0, 0, $width, $height, "RGBA", Imagick::PIXEL_CHAR);`

The crash was happens because of RAM limitations. Before calling exportImagePixels script use around 2 mb of RAM, after using export it use 258 mb of RAM(parameter $width=3000, $heignt = 500). I suppose that it should be around 6 mb for this parameters. For example for exported image with size 300x300 it use 18 mb that is a lot.

For measuring memory I use following function:

function printMemoryUsage()  {
        $mem_usage = memory_get_usage(true);
        if ($mem_usage < 1024)
            echo $mem_usage." bytes";
        elseif ($mem_usage < 1048576)
            echo round($mem_usage/1024,2)." kilobytes";
        else
            echo round($mem_usage/1048576,2)." megabytes";
}
Do you have any idea why it use such amount of memories, and how I could just get a pixel array from Imagick image?

I use PHP version: 7.4.33
ImageMagick 6.9.10-68 Q16 x86_64 2021-10-14

**Comment by snibgo** [comment]
What is the text output from "convert -version"?

Your processing using IMagick, which is a PHP interface to ImageMagick, and is not developed or supported by the ImageMagick people. This forum is for ImageMagick, not IMagick.

> Before calling exportImagePixels script use around 2 mb of RAM, ...

That seems unlikely. Your image is 3000x2000 pixels, with four channels. If your v6 IM is Q8 without HDRI, this needs 3000x2000x4 = 24 MB.

**Comment by retsetB** [comment]
convert -version
Version: ImageMagick 6.9.10-68 Q16 x86_64 2021-10-14 https://imagemagick.org
Copyright: © 1999-2019 ImageMagick Studio LLC
License: https://imagemagick.org/script/license.php
Features: Cipher DPC Modules OpenMP(3.1)
Delegates (built-in): bzlib cairo fontconfig freetype gslib jng jp2 jpeg lcms ltdl lzma openexr pangocairo png ps rsvg tiff wmf x xml zlib

about 2 mb I also was surprised, but I use code for checking that is kind of correct

**Comment by urban-warrior** [comment]
Image processing can be memory intensive.   Typically you would increase the PHP memory limits in `php.ini` to accommodate the demands of ImageMagick.  If you cannot, you can force ImageMagick to shuffle pixels to/from disk rather than memory with the requisite penalty in performance.  Try setting the memory and map limits, or an easier way is to set the single area limit-- which forces any image over the area limit to cache pixels on disk rather than memory.  See https://imagemagick.org/script/security-policy.php to set the area limit as a policy or https://imagemagick.org/script/resources.php to set it as an environment variable.   We don't use iMagick, but its likely there is a supported method to set limits right in the iMagick extension.
